substrate/network/src/on_demand.rs

```diff
@@ -0,0 +1,374 @@
+		let mut core = self.core.lock();
+		for bad_peer in core.maintain_peers() {
+			trace!(target: "sync", "Remote request timeout for peer {}", bad_peer);
+			io.disconnect_peer(bad_peer);
+		}
+		core.dispatch();
+	}
+
+	fn on_remote_response(&self, io: &mut SyncIo, peer: PeerId, response: message::RemoteCallResponse) {
+		let mut core = self.core.lock();
+		match core.remove(peer, response.id) {
+			Some(request) => {
+				// we do not bother if receiver has been dropped already
+				let _ = request.sender.send(response);
+			},
+			None => {
+				trace!(target: "sync", "Invalid remote response from peer {}", peer);
+				io.disconnect_peer(peer);
+				core.remove_peer(peer);
+			},
+		}
+
+		core.dispatch();
+	}
+}
+
+impl<E> Fetcher for OnDemand<E> where E: service::ExecuteInContext {
+	fn execution_proof(&self, block: HeaderHash, method: &str, call_data: &[u8]) -> client::error::Result<(Vec<u8>, Vec<Vec<u8>>)> {
+		self.remote_call(block, method, call_data).wait().map_err(|_| unimplemented!())
+	}
+}
+
+impl<E> OnDemandCore<E> where E: service::ExecuteInContext {
+	pub fn add_peer(&mut self, peer: PeerId) {
+		self.idle_peers.push_back(peer);
+	}
+
+	pub fn remove_peer(&mut self, peer: PeerId) {
+		if let Some(request) = self.active_peers.remove(&peer) {
+			self.pending_requests.push_front(request);
+			return;
+		}
+
+		if let Some(idle_index) = self.idle_peers.iter().position(|i| *i == peer) {
+			self.idle_peers.swap_remove_back(idle_index);
+		}
+	}
+
+	pub fn maintain_peers(&mut self) -> Vec<PeerId> {
+		let now = Instant::now();
+		let mut bad_peers = Vec::new();
+		loop {
+			match self.active_peers.front() {
+				Some((_, request)) if now - request.timestamp >= REQUEST_TIMEOUT => (),
+				_ => return bad_peers,
+			}
+
+			let (bad_peer, request) = self.active_peers.pop_front().expect("front() is Some as checked above");
+			self.pending_requests.push_front(request);
+			bad_peers.push(bad_peer);
+		}
+	}
+
+	pub fn insert(&mut self, sender: Sender<message::RemoteCallResponse>, mut message: message::RemoteCallRequest) {
+		message.id = self.next_request_id;
+		self.next_request_id += 1;
+		self.pending_requests.push_back(Request {
+			id: message.id,
+			timestamp: Instant::now(),
+			sender,
+			message,
+		});
+	}
+
+	pub fn remove(&mut self, peer: PeerId, id: u64) -> Option<Request> {
+		match self.active_peers.entry(peer) {
+			Entry::Occupied(entry) => match entry.get().id == id {
+				true => {
+					self.idle_peers.push_back(peer);
+					Some(entry.remove())
+				},
+				false => None,
+			},
+			Entry::Vacant(_) => None,
+		}
+	}
+
+	pub fn dispatch(&mut self) {
+		let service = match self.service.upgrade() {
+			Some(service) => service,
+			None => return,
+		};
+
+		while !self.pending_requests.is_empty() {
+			let peer = match self.idle_peers.pop_front() {
+				Some(peer) => peer,
+				None => return,
+			};
+
+			let mut request = self.pending_requests.pop_front().expect("checked in loop condition; qed");
```

Comment:
can't we just loop `for mut x in pending_requests`?

Reply:
there might be no peers available or number of peers is < than number of requests (currently 1 request <-> 1 peer at a time)